### PR TITLE
ADD PHP SUPPORT FOR LICENSE CRAWLING

### PR DESCRIPTION
**Pull request by Fredrick-Tam**
This PR is to address issue https://github.com/clearlydefined/crawler/issues/230 by starting to enable php license crawling.

Using https://github.com/clearlydefined/crawler/pull/110/ and https://github.com/clearlydefined/crawler/pull/86 as guides

### CODE REVIEW

File: providers/fetch/packagistFetch.js [hunk 1]
```diff
@@ -0,0 +1,80 @@
+// Copyright (c) Microsoft Corporation and others. Licensed under the MIT license.
+// SPDX-License-Identifier: MIT
+
+const AbstractFetch = require('./abstractFetch')
+const requestRetry = require('requestretry').defaults({ maxAttempts: 3, fullResponse: true })
+const fs = require('fs')
+const request = require('request')
+// const requestPromise = require('request-promise-native')
```
Comment (geneh) [inline providers/fetch/packagistFetch.js]: Should this line be removed?

File: providers/fetch/packagistFetch.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    registryData.releaseDate = registryData.manifest['time']
+    delete registryData['packages']
+    return registryData
+  }
+
+  async _getPackage(spec, registryData, destination) {
+    return new Promise((resolve, reject) => {
+      var options = {
```
Comment (geneh) [inline providers/fetch/packagistFetch.js]: It should be const instead of var.

File: providers/fetch/packagistFetch.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        url: registryData.manifest.dist.url,
```
Comment (geneh) [inline providers/fetch/packagistFetch.js]: Can dist be undefined? Perhaps lodash get(...) is safer?

File: providers/process/composerExtract.js [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+const { get, merge } = require('lodash')
+
+class ComposerExtract extends AbstractClearlyDefinedProcessor {
+  constructor(options, sourceFinder) {
+    super(options)
+    this.sourceFinder = sourceFinder
+  }
+
+  get toolVersion() {
+    return '1.1.4'
```
Comment (geneh) [inline providers/process/composerExtract.js]: It should probably be 1.0.0

File: providers/process/composerExtract.js [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+  async handle(request) {
+    // skip all the hard work if we are just traversing.
+    if (this.isProcessing(request)) {
+      const location = request.document.location
+      await super.handle(request, location, 'package')
+      const manifestLocation = this._getManifestLocation(location)
+      const manifest = manifestLocation ? JSON.parse(fs.readFileSync(path.join(location, manifestLocation))) : null
+      await this._createDocument(request, manifest, request.document.registryData)
+      if (manifest) this.attachFiles(request.document, [manifestLocation], location)
+      else console.log('PHP package without composer.json', { url: request.url })
```
Comment (geneh) [inline providers/process/composerExtract.js]: console.log should be replaced with logger.info

File: test/unit/providers/fetch/packagistFetchTests.js [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+const expect = require('chai').expect
+const sinon = require('sinon')
+const PassThrough = require('stream').PassThrough
+const PackagistFetch = require('../../../../providers/fetch/packagistFetch')
+const proxyquire = require('proxyquire')
+const Request = require('ghcrawler').request
+const fs = require('fs')
+
+it('can handle the request being attempted', async () => {
```
Comment (geneh) [inline test/unit/providers/fetch/packagistFetchTests.js]: Should this test be inside describe?

File: test/unit/providers/fetch/packagistFetchTests.js [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+let Fetch
+
+const hashes = {
+  'symfony-polyfill-mbstring-v1.11.0-0-gfe5e94c.zip': {
+    sha1: '8d24c52e593042529ba86549d9920eb4d9649763',
+    sha256: '797a607b7ea7dad62f78a56f3687f2b2108d221b0682d0ea1386db61714dc8a2'
+  }
+}
+
+describe('', () => {
```
Comment (geneh) [inline test/unit/providers/fetch/packagistFetchTests.js]: Let's add description text here.

File: test/unit/providers/fetch/packagistFetchTests.js [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+  beforeEach(() => {
+    const resultBox = {}
+    const requestPromiseStub = options => {
+      if (options.url) {
+        if (options.url.includes('regError')) throw new Error('yikes')
```
Comment (geneh) [inline test/unit/providers/fetch/packagistFetchTests.js]: I would replace 'yikes' with something more meaningful

File: providers/process/composerExtract.js (hunk 5 above)
Comment (Fredrick-Tam) [reply providers/process/composerExtract.js]: For some weird reason, I cant call `this.logger.info` here, throws an error

Comment (Fredrick-Tam) [reply providers/process/composerExtract.js]: figured it out

**Comment by Fredrick-Tam** [comment]
🙇 awesome 💯

**Pull request merged**